File: 백준/Silver/17503. 맥주 축제/맥주 축제.cc
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define MOD 1000000007
#define MAX 100000
#define INF 987654321
ll dx[] = {1, -1, 0, 0};
ll dy[] = {0, 0, 1, -1};
ll n, m, k;
vector<pair<ll, ll>> v;

bool check(ll target)
{
    ll sum = 0, cnt = 0;
    for (ll i = 0; i < k; i++)
    {
        if (v[i].second <= target)
        {
            sum += v[i].first;
            cnt++;
        }
        if (cnt >= n)
            break;
    }
    if (cnt < n)
        return false;
    return sum >= m;
}

ll bin_search()
{
    ll lo = 0, hi = 1e12;

    while (lo + 1 < hi)
    {
        ll mid = (lo + hi) / 2;

        if (!check(mid))
            lo = mid;
        else
            hi = mid;
    }

    return hi;
}
int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    cin >> n >> m >> k;

    for (ll i = 0; i < k; i++)
    {
        pair<ll, ll> a;
        cin >> a.first >> a.second;
        v.push_back(a);
    }

    sort(v.begin(), v.end(), [](pair<ll, ll> &a, pair<ll, ll> &b)
         { return a.first > b.first; });

    ll res = bin_search();

    if (res == 1e12)
        cout << "-1";
    else
        cout << res;
    return 0;
}

File: 백준/Silver/17503. 맥주 축제/맥주 축제.py
import sys

input = sys.stdin.readline
print = sys.stdout.write


def check(target):
    sum = 0
    cnt = 0
    for a in v:
        if a[1] <= target:
            sum += a[0]
            cnt += 1

        if cnt == n:
            break
    return sum >= m if cnt == n else False


def bin_search():
    lo, hi = 0, int(1e12)

    while lo + 1 < hi:
        mid = (lo + hi) // 2
        if check(mid):
            hi = mid
        else:
            lo = mid
    return hi


n, m, k = map(int, input().split())
v = [tuple(map(int, input().split())) for _ in range(k)]
v = sorted(v, key=lambda x: -x[0])

if (res := bin_search()) != int(1e12):
    print(str(res))
else:
    print("-1")
